x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -0,0 +1,263 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.inference.results.TextExpansionResults.WeightedToken;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.ml.queries.TokenPruningConfig.PRUNING_CONFIG;
+
+public class WeightedTokensQueryBuilder extends AbstractQueryBuilder<WeightedTokensQueryBuilder> {
+    public static final String NAME = "weighted_tokens";
+
+    public static final ParseField TOKENS_FIELD = new ParseField("tokens");
+    private final String fieldName;
+    private final Set<WeightedToken> tokens;
+    @Nullable
+    private final TokenPruningConfig tokenPruningConfig;
+
+    public WeightedTokensQueryBuilder(String fieldName, Set<WeightedToken> tokens) {
+        this(fieldName, tokens, null);
+    }
+
+    public WeightedTokensQueryBuilder(String fieldName, Set<WeightedToken> tokens, @Nullable TokenPruningConfig tokenPruningConfig) {
+        this.fieldName = Objects.requireNonNull(fieldName, "[" + NAME + "] requires a fieldName");
+        this.tokens = Objects.requireNonNull(tokens, "[" + NAME + "] requires tokens");
+        if (tokens.isEmpty()) {
+            throw new IllegalArgumentException("[" + NAME + "] requires at least one token");
+        }
+        this.tokenPruningConfig = tokenPruningConfig;
+    }
+
+    public WeightedTokensQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.tokens = in.readCollectionAsSet(WeightedToken::new);
+        this.tokenPruningConfig = in.readOptionalWriteable(TokenPruningConfig::new);
+    }
+
+    public String getFieldName() {
+        return fieldName;
+    }
+
+    @Nullable
+    public TokenPruningConfig getTokenPruningConfig() {
+        return tokenPruningConfig;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeCollection(tokens);
+        out.writeOptionalWriteable(tokenPruningConfig);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.startObject(TOKENS_FIELD.getPreferredName());
+        for (var token : tokens) {
+            token.toXContent(builder, params);
+        }
+        builder.endObject();
+        if (tokenPruningConfig != null) {
+            tokenPruningConfig.toXContent(builder, params);
+        }
```

Comment:
Again, this builder should determine the field name and the pruning config just writes its object and does not dictate the field name.